File: julia/003. largest prime factor.jl
# Largest prime factor
# https://projecteuler.net/problem=3

function is_prime(n)
    if n != 2 && n % 2 == 0 || n == 0 || n == 1
        return false
    end
    for i in 3:2:sqrt(n)
        if n % i == 0
            return false
        end
    end
    return true
end

function largest_prime_factor(number)
    if number == 0 || number == 1
        return "n/a"
    end
    if is_prime(number)
        return number
    end
    prime_factors = []
    for i in 2:floor(sqrt(number))
        if number % i != 0
            continue
        end
        if is_prime(i)
            push!(prime_factors, floor(Int, i))
        end
        if is_prime(number / i)
            push!(prime_factors, floor(Int, number / i))
        end
    end
    return maximum(prime_factors)
end

println(largest_prime_factor(600851475143))

File: javascript/003_largest_prime_factor.js
// Largest prime factor
// https://projecteuler.net/problem=3

function isPrime(n) {
    if ((n % 2 == 0 && n != 2) || n == 0 || n == 1) {
        return false;
    }
    for (let i = 3; i <= Math.sqrt(n); i += 2) {
        if (n % i == 0) {
            return false;
        }
    }
    return true;
}

function largestPrimeFactor(number) {
    if (number == 0 || number == 1 || number != Math.floor(number)) {
        return "n/a";
    }
    if (isPrime(number)) {
        return number;
    }
    const primeFactors = [];
    for (let i = 2; i <= Math.sqrt(number); i++) {
        if (number % i != 0) {
            continue;
        }
        if (isPrime(i)) {
            primeFactors.push(i);
        }
        if (isPrime(number / i)) {
            primeFactors.push(number / i);
        }
    }
    primeFactors.sort((a, b) => a - b);
    return primeFactors.at(-1);
}

console.log(largestPrimeFactor(600851475143));
